addons/sale/models/payment_transaction.py

```diff
@@ -46,34 +53,37 @@ def _set_pending(self, state_message=None):
             sales_orders._send_order_confirmation_mail()
 
     def _check_amount_and_confirm_order(self):
-        self.ensure_one()
-        for order in self.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent')):
-            if order.currency_id.compare_amounts(self.amount, order.amount_total) == 0:
-                order.with_context(send_email=True).action_confirm()
-            else:
-                _logger.warning(
-                    '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %r, got %r',
-                    self.acquirer_id.provider,order.name, order.id,
-                    order.amount_total, self.amount,
+        """ Auto-confirm sales orders based on the amount of all the transactions linked. """
+        for tx in self:
+            for order in tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent')):
+                tx_total_amount = math.fsum(  # Sum the amount of each transaction in the SO
+                    order.transaction_ids.mapped(lambda tx: float(tx.amount))
                 )
```

Comment:
Missing check on currencies, what if the tx currency is different than the order currency ? (Not your fault, it wasn't verified in old code).

Reply:
I didn't know that was possible. In that case, it means that I need to convert on an arbitrary currency using the current rate? But what if the current rate change between the moment when the transaction is made and the moment that the post-process is made? Do we have a history of rates so I can fetch the current rate at the time when the transaction was paid?

Reply:
It's a tricky case that may happen (but shouldn't indeed :D).
About currency rates, there exists tools to compute rates at a given date (and records to store such info), but I don't think you should go that far.

If the transaction currencies are different, you should skip SO validation IMHO, we don't want to support that case (I'm not even sure we want to support the confirmation of the SO after multiple partial payments ^^).